subsys/storage/flash_map/flash_map_default.c

```diff
@@ -19,26 +17,48 @@
 	 .fa_dev = DEVICE_DT_GET(DT_MTD_FROM_FIXED_PARTITION(part)),	        \
 	 .fa_size = DT_REG_SIZE(part),						\
 	 .fa_label = DT_PROP_OR(part, label, NULL),	},
+#define FLASH_AREA_SUB_FOO(part)						\
+	{.fa_id = DT_FIXED_PARTITION_ID(part),					\
+	 .fa_off = DT_REG_ADDR(part),						\
+	 .fa_dev = DEVICE_DT_GET(DT_MTD_FROM_FIXED_SUBPARTITION(part)),		\
+	 .fa_size = DT_REG_SIZE(part),						\
+	 .fa_label = DT_PROP_OR(part, label, NULL),	},
 #else
 #define FLASH_AREA_FOO(part)							\
 	{.fa_id = DT_FIXED_PARTITION_ID(part),					\
 	 .fa_off = DT_REG_ADDR(part),						\
 	 .fa_dev = DEVICE_DT_GET(DT_MTD_FROM_FIXED_PARTITION(part)),	        \
 	 .fa_size = DT_REG_SIZE(part), },
+#define FLASH_AREA_SUB_FOO(part)						\
+	{.fa_id = DT_FIXED_PARTITION_ID(part),					\
+	 .fa_off = DT_REG_ADDR(part),						\
+	 .fa_dev = DEVICE_DT_GET(DT_MTD_FROM_FIXED_SUBPARTITION(part)),		\
+	 .fa_size = DT_REG_SIZE(part), },
 #endif
 
+
 #define FLASH_AREA_FOOO(part)	\
 	COND_CODE_1(DT_NODE_HAS_STATUS_OKAY(DT_MTD_FROM_FIXED_PARTITION(part)), \
 		(FLASH_AREA_FOO(part)), ())
 
 #define FOREACH_PARTITION(n) DT_FOREACH_CHILD(DT_DRV_INST(n), FLASH_AREA_FOOO)
 
+#define FLASH_AREA_SUB_FOOO(part)	\
+	COND_CODE_1(DT_NODE_HAS_STATUS_OKAY(DT_MTD_FROM_FIXED_SUBPARTITION(part)), \
+		(FLASH_AREA_SUB_FOO(part)), ())
+
+#define FOREACH_SUB_PARTITION(n) DT_FOREACH_CHILD(DT_DRV_INST(n), FLASH_AREA_SUB_FOOO)
+
 /* We iterate over all compatible 'fixed-partitions' nodes and
  * use DT_FOREACH_CHILD to iterate over all the partitions for that
  * 'fixed-partitions' node.  This way we build a global partition map
  */
 const struct flash_area default_flash_map[] = {
+#define DT_DRV_COMPAT fixed_partitions
```

Comment:
Should be able to use `DT_FOREACH_STATUS_OKAY` directly to avoid all the define hackery
```
   DT_FOREACH_STATUS_OKAY(fixed_partitions, FOREACH_PARTITION)
   DT_FOREACH_STATUS_OKAY(fixed_subpartitions, FOREACH_SUB_PARTITION)
```